api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -135,9 +135,12 @@ message Cluster {
   }
 
   enum ClusterProtocolSelection {
-    // Cluster can only operate on one of the possible upstream protocols (HTTP1.1, HTTP2).
-    // If :ref:`http2_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http2_protocol_options>` are
-    // present, HTTP2 will be used, otherwise HTTP1.1 will be used.
+    // If both :ref:`http2_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http2_protocol_options>`
+    // and :ref:`http_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http_protocol_options>` are
+    // configured, Envoy will attempt to do ALPN negotiation for TLS connections, failing
+    // over to HTTP/1.1 if ALPN negotiation fails.
+    // If only one protocol option is present it will be used as the hard-coded
+    // protocol. If neither is present, HTTP/1.1 will be used.
```

Comment:
A constant usability issue for users is they get confused about how to actually configure the appropriate ALPN. A few thoughts:
1) In a prefect world, would we have a new protocol selection mode about H2_H1_ALPN which force configures the ALPN configuration correctly (or at least checks it) for "h2,http/1.1" ?
2) If we can't do force setting, or even checking, due to the transport protocol extensibility (and risk of the ALPN values being not what people want for some reason), can we beef up the documentation here quite a bit with ref linking, examples, etc. on how to properly set ALPN for the common case?

cc @phlax also who has been doing a lot of great work on TLS sandboxes and examples.

Reply:
Most of this PR is about the code once we've decided to do ALPN, so I think we can review that in parallel if you have the cycles, but let's talk API regardless :-)

I don't like the idea of H1_H2 ALPN because I think for mobile we're going to want H3_H2_H1 by default (using the highest available) and I want to plan for doing H3 where abvailable.

Ideas offhand

We could leave this as-is, and beef up comments.  When we have H3-upstream folks can just add that in for the happy-eyeballs style "try all, use best of"

We could leave USE_CONFIGURED_PROTOCOL as "http1 OR http2" and add USE_ALL_CONFIGURED_PROTOCOLS for alpn, so there's explicit config.  Maybe reject USE_ALL if there's only one protocol configured?     I don't think there's much of a gain here, except rejecting if only one protocol.

We could add USE_LISTED_PROTOCOLS and an enum which is only valid iff USE_LISTED_PROTOCOLS is set, which currently allows for HTTP/1 and/or HTTP/2 (and could eventually be extended for HTTP/3 probing and use). We could just add this or add this and deprecate USE_CONFIGURED_PROTOCOLS.   

We could deprecate this whole section and do something cleaner (suggestions welcome)

Reply:
cc @envoyproxy/api-shepherds 

To be perfectly honest the `Cluster` API is a raging disaster. It has all kinds of fields that are protocol/extension specific and should not be there. In a perfect world most of the messages currently defined in `Cluster` would move out into extension configuration and then be loaded as part of opaque configuration depending on the protocols being used. E.g., you can imagine all of the LB configurations being opaque, protocol configurations being opaque, etc.

With all of that said, I really doubt there is going to be enough inertia to clean this up, especially within a minor API version.

So, given the constraints of the world we actually live in, what can we do? I think there are a few reasonable options, depending on how much you are willing to take on. Here is what I would suggest which I think balances not creating more mess with something that isn't too drastic.

1) Deprecate ClusterProtocolSelection, protocol_selection, common_http_protocol_options, http_protocol_options, and http2_protocol_options
2) Move HTTP specific protocol configuration into a new typed configuration driven through typed_extension_protocol_options.
3) Create a new wrapper message which is something like ClusterHttpProtocolConfiguration.
4) Inside this message, defined a oneof which allows specific types of configuration.
5) Add an ExplicitHttpConfig message which has has common HTTP options, and a oneof between H1 and H2 (and later H3)
6) Add a UseDownstreamHttpConfig option with a similar message layout as (5).
7) Add a AlpnHttpConfig option which has common options and then no oneof around each protocol.
8) For the ALPN option, I was going to suggest what you suggested below, is add ALPN inspection to the transport socket interface which allows for checking what is configured so that we can actually sensibly warn if nothing is set.

^ is probably more work than you want to do, but if you are willing to take one for the team we can slay a lot of tech debt here. Obviously there will be details to sort out in the final API, but let me know what you think.

Reply:
I think that 6 should have the same layout as 7) since you can configure H1 and H2 and use whatever downstream dows.

So something like

  message ExplicitHttpConfig {
    core.v3.HttpProtocolOptions common_http_protocol_options = 1;

    oneof SingleProtocolConfig {
      core.v3.Http1ProtocolOptions http_protocol_options = 1;
      core.v3.Http1ProtocolOptions http2_protocol_options = 2;
    }
    SingleProtocolConfig protocol_config = 2;
  };

  message UseDownstreamHttpConfig {
    core.v3.HttpProtocolOptions common_http_protocol_options = 1;
    core.v3.Http1ProtocolOptions http_protocol_options = 1;
    core.v3.Http1ProtocolOptions http2_protocol_options = 2;
  };

  message AlpnHttpConfig {
    core.v3.HttpProtocolOptions common_http_protocol_options = 1;
    core.v3.Http1ProtocolOptions http_protocol_options = 1;
    core.v3.Http1ProtocolOptions http2_protocol_options = 2;
  };

Reply:
Yeah pretty much. Do you want to take a crack at the API changes and we can just review that?

Reply:
Sent!  I didn't run CI checks or do the Envoy integration, so please ignore any CI failures while we sort out the API plan.